doc/neps/nep-0055-string_dtype.rst

```diff
@@ -656,13 +518,385 @@ string validation in a separate, expensive, pass over the input array-like. We
 have chosen not to make this the default behavior to follow NumPy fixed-width
 strings, which coerce non-strings.
 
+C API for ``StringDType``
+*************************
+
+Each entry in a ``StringDType`` array will be defined to be an instance of an
+opaque ``npy_packed_static_string`` struct. Making the public definition opaque
+will allow us to in principle change the underlying C implementation or even the
+size of the struct in the future.
+
+To actually work with the string data, we will publicly expose a typedef for
+a struct with with the following definition:
+
+.. code-block:: C
+
+   struct npy_static_string {
+       size_t size;
+       const char *buf;
+   };
+
+Where ``size`` is the size, in bytes, of the string and ``buf`` is a const
+pointer to the beginning of a UTF-8 encoded bytestream containing string
+data. This is a *read-only* view onto the string, we will not expose a public
+interface for modifying these strings. We do not append a trailing null
+character to the byte stream, so users attempting to pass the ``buf`` field to
+an API expecting a C string must create a copy with a trailing null. As a
+positive consequence, ``StringDType`` array entries can contain arbitrary
+embedded or trailing null characters.
```

Comment:
I am happy with this, and we can always consider to change it in some future also if it proofs annoying.  In practice, sometimes a trailing null is nice (even when allowing embedded ones).

Reply:
Added a sentence explaining that we can choose to add the trailing null later if we want.